### PR TITLE
Feature: Added Dockerfile to Create Docker Image Of Pynecone

### PR DESCRIPTION
**Pull request by SaarthakMaini**
This PR involves:

* Adding a Dockerfile
* Building Docker Image For Pynecone

To Run The Container Locally Do The Following:

*  Install Docker On Your Local Machine
*  Clone the pynecone repository by running the command: `git clone https://github.com/pynecone-io/pynecone.git`
*  RUN:  `docker build -t pynecone .` inside the Dockerfile Directory of the terminal
*  Then RUN:  `docker run -it -d -p 3000:3000 pynecone_image`
*  After that you can see the Demo Webpage on `http://localhost:3000` OR we can view the content by typing the following command in the terminal : `curl http://localhost:3000`

I would like to give the suggestion of pushing the Docker Image to DockerHub

### CODE REVIEW

**Comment by r0b2g1t** [comment]
What do you think of the idea of building a multistage image? This would dramatically reduce the size of the container image.

**Comment by r0b2g1t** [comment]
I would recommend to create a docerfile sample with which every Pynecone developer can build an optimised container image himself. To my eye, this Dockerfile produces an image that only produces a sample app and cannot be used for other projects. For this reason, I would not recommend pushing the resulting image to Docker Hub or any other public registry.

File: Dockerfile/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,30 @@
+FROM python:3.8-slim-buster
+
+RUN pip install --upgrade pip
+
+RUN apt-get update
+
+RUN apt-get install curl unzip --yes
+
+RUN apt-get update && apt-get install -y nodejs npm
```
Comment (r0b2g1t) [inline Dockerfile/Dockerfile:9]: Add apt cache clean-up afterwards, it will shrink the image size.

File: Dockerfile/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+RUN curl -fsSL https://deb.nodesource.com/setup_16.x | bash - && apt-get install -y nodejs
```
Comment (r0b2g1t) [inline Dockerfile/Dockerfile:10]: Change to version 18 the latest LTS release. For 16 the active support ended 3 months ago and security patches will only provided for the next 7 months.

File: Dockerfile/Dockerfile (hunk 1 above)
Comment (r0b2g1t) [inline Dockerfile/Dockerfile:9]: Why do you install NodeJS twice first in line 9 and again with a another way in line 10?

Comment (r0b2g1t) [inline Dockerfile/Dockerfile:9]: Lines 5-9 can be combined to:
`RUN apt-get update && apt-get install -y curl nodejs npm unzip`

**Comment by FHU-yezi** [comment]
In my option, we should combine RUN lines to single command, this result less layers and snaller image size.

File: Dockerfile/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,30 @@
+FROM python:3.8-slim-buster
+
+RUN pip install --upgrade pip
```
Comment (r0b2g1t) [inline Dockerfile/Dockerfile:3]: can removed you should use a latest image where the latest python and pip version is included.

File: Dockerfile/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+RUN mkdir demo_pynecone_app
+
+WORKDIR /pynecone/demo_pynecone_app
+
+RUN pc init
+
+CMD ["pc","run"]
+
+EXPOSE 3000
+EXPOSE 8000
```
Comment (r0b2g1t) [inline Dockerfile/Dockerfile:29]: Why do you expose port 8000?

**Review by r0b2g1t — CHANGES_REQUESTED** [review]
Discussion is needed to clarify the finding.

File: Dockerfile/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+RUN apt-get update && apt-get install -y nodejs npm
+RUN curl -fsSL https://deb.nodesource.com/setup_16.x | bash - && apt-get install -y nodejs
+
+RUN python3 -m venv /opt/venv
+
+RUN . /opt/venv/bin/activate
+
+RUN pip install pynecone-io
+
+WORKDIR /pynecone
```
Comment (r0b2g1t) [inline Dockerfile/Dockerfile:18]: Obsolete by line 22 - `WORKDIR` will create subdirectory.

File: Dockerfile/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+RUN mkdir demo_pynecone_app
```
Comment (r0b2g1t) [inline Dockerfile/Dockerfile:20]: Obsolete by line 22 - `WORKDIR` will create subdirectory.

**Review by r0b2g1t — CHANGES_REQUESTED** [review]

File: Dockerfile/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,30 @@
+FROM python:3.8-slim-buster
```
[the remark this answers is not in the document]
Comment (picklelo) [reply Dockerfile/Dockerfile:1]: Agreed, let's use the latest Python

File: Dockerfile/Dockerfile (hunk 1 above)
Comment (picklelo) [reply Dockerfile/Dockerfile:9]: We can also combine these into a single RUN command to reduce the number of steps:
```
RUN apt-get update; \
    apt-get install curl unzip --yes \
    apt-get update && apt-get install -y nodejs npm
```

File: Dockerfile/Dockerfile (hunk 2 above)
Comment (picklelo) [reply Dockerfile/Dockerfile:10]: +1 The apt-get should already be enough

File: Dockerfile/Dockerfile (hunk 4 above)
Comment (picklelo) [reply Dockerfile/Dockerfile:29]: Port 8000 is the backend FastAPI port, while the NextJS app is on port 3000. So we need the 8000 port to connect to the websocket

File: Dockerfile/Dockerfile (hunk 7 above)
[the remark this answers is not in the document]
Comment (SaarthakMaini) [reply Dockerfile/Dockerfile:1]: Thank you very much for the feedback!😁

I will push all the changes at once to update the PR, till then I would be updating the discussion as Done if I have solved the corresponding issue on my local machine

Updated `python:3.8-slim-buster` to `python:3.11.1-slim-buster`

File: Dockerfile/Dockerfile (hunk 3 above)
Comment (SaarthakMaini) [reply Dockerfile/Dockerfile:3]: Thank you!
I have deleted this line and it's working perfectly!

File: Dockerfile/Dockerfile (hunk 1 above)
Comment (SaarthakMaini) [reply Dockerfile/Dockerfile:9]: Solved this issue locally, will update PR once rest of the issues are resolved

File: Dockerfile/Dockerfile (hunk 6 above)
Comment (SaarthakMaini) [reply Dockerfile/Dockerfile:20]: Solved

File: Dockerfile/Dockerfile (hunk 5 above)
Comment (SaarthakMaini) [reply Dockerfile/Dockerfile:18]: Solved

File: Dockerfile/Dockerfile (hunk 1 above)
Comment (SaarthakMaini) [reply Dockerfile/Dockerfile:9]: @r0b2g1t  @picklelo  Thank you for your feedback! I have done most of the changes as suggested, but I'm facing the following problems:

- When I am trying to install nodejs via `RUN apt-get update && apt-get install -y curl nodejs npm unzip` , my script is hanging and docker stops working completely. To make it work again I have to restart my computer and check my Hyper-V settings. The same problem is occuring when I am trying to install Version 18 of Nodejs

- apt cache clean-up is not running and giving error while running the script

If possible, could you please look into it and direct me towards a possible solution?

Thank You!😁

Comment (r0b2g1t) [reply Dockerfile/Dockerfile:9]: Hey @SaarthakMaini sure I will look to this issue and try it out and give you feedback. 🙂

**Comment by r0b2g1t** [comment]
I have open another PR #387 with all the changes. There were too many changes that I should have described here.

**Comment by r0b2g1t** [comment]
This PR could be closed #387 is merged already.

**Comment by SaarthakMaini** [comment]
The PR looks awesome @r0b2g1t !

If possible, I would love to get some feedback on how I can improve my skills and what I could have done better.😃

Thank You!

**Comment by Alek99** [comment]
> The PR looks awesome @r0b2g1t !
> 
> If possible, I would love to get some feedback on how I can improve my skills and what I could have done better.😃
> 
> Thank You!

Thanks for working on that last PR Saarthak we really appreciate It. Many more issues we'll need help on to come and can definitely use your help. Sorry we were not able to merge in your pr we just accepted the other pr because it was finished and we were trying to get the release out by today. 

We can sync on discord soon and decide more things to work on for the next release if you have time. 🙂

**Pull request closed without merging**